File: 3kurs/2sem/paralelki/stsala/lab2/src/main/scala/CpuQueue.scala
import scala.collection.mutable

class CpuQueue(val capacity: Int, val threadName: String) {
  private val queue = new mutable.Queue[String]()
  var maxSize = 0
  
  def push(process: String): Unit = this.synchronized {
    while (queue.size == capacity) {
      println(s"queue ${threadName} is full, waiting")
      wait()
    }
    queue += process
    if (maxSize < queue.size) maxSize = queue.size
    
    println(s"process ${process} added")
    println(s"queue ${threadName} : ${queue}")
    notify()
  }
  
  def pop(secondQueue: CpuQueue, secondaryChecked: Boolean = false): String = {
    var checked = secondaryChecked
    this.synchronized {
      if (secondaryChecked && isEmpty) {
        checked = false
        wait()
      }
      if (!isEmpty) {
        val res = queue.dequeue()
        notify()
        return res
      }
    }
    
    println(s"queue ${threadName} is empty, checking secondary")
    
    secondQueue.synchronized {
      if (!secondQueue.isEmpty) {
        println(s"cpu ${threadName} getting element from secondary queue")
        val res = secondQueue.queue.dequeue()
        secondQueue.notify()
        return res
      }
    }
    checked = true
    println(s"both queues empty, waiting")
    pop(secondQueue, checked)
  }
  
  def isEmpty: Boolean = {
    queue.isEmpty
  }
}

object CpuQueue {
  def apply(capacity: Int, threadName: String): CpuQueue = new CpuQueue(capacity, threadName)
}

File: 3kurs/2sem/paralelki/jaba/lab2/src/main/java/CpuQueue.java
import java.util.*;

public class CpuQueue {
    
    private Queue<Integer> queue = new LinkedList<>();
    private int capacity;
    private int index;
    private int maxSize = 0;
    
    public CpuQueue(int capacity, int index) {
        this.capacity = capacity;
        this.index = index;
    }
    
    public int getMaxSize() {
        return maxSize;
    }
    
    public synchronized int get() throws InterruptedException {
        while (queue.isEmpty()) {
            System.out.println("Queue is EMPTY, waiting..");
            this.wait();
        }
        int item = queue.remove();
        System.out.println("Process removed, queue size = [" + queue.size() + "]");
        return item;
    }
    
    public synchronized boolean put(int element) {
        if (this.queue.size() == capacity) {
            System.out.println("Queue " + index + " is FULL");
            return false;
        }
        this.queue.add(element);
        
        if (this.queue.size() > maxSize) {
            maxSize = this.queue.size();
        }
        
        System.out.println("Process added, queue [" + index + "] size = [" + this.queue.size() + "]\n");
        this.notify();
        return true;
    }
}
